noir/aztec_macros/src/lib.rs

```diff
@@ -1601,3 +1601,143 @@ fn event_signature(event: &StructType) -> String {
     let fields = vecmap(event.get_fields(&[]), |(_, typ)| signature_of_type(&typ));
     format!("{}({})", event.name.0.contents, fields.join(","))
 }
+
+fn inject_compute_note_hash_and_nullifier(
+    crate_id: &CrateId,
+    context: &mut HirContext,
+    unresolved_traits_impls: &Vec<UnresolvedTraitImpl>,
+    collected_functions: &mut Vec<UnresolvedFunctions>,
+) -> Result<(), (MacroError, FileId)> {
+    // We first fetch modules in this crate which correspond to contracts, along with their file id.
+    let mut contract_module_file_ids: Vec<(LocalModuleId, FileId)> = context.def_map(crate_id).unwrap().modules().iter()
+        .filter(|(_, module)| module.is_contract)
+        .map(|(idx, module)| (LocalModuleId(idx), module.location.file))
+        .collect();
+
+    // If the current crate does not contain a contract module we simply skip it.
+    if contract_module_file_ids.len() == 0 {
+        return Ok(());
+    } else if contract_module_file_ids.len() != 1 {
+        panic!("Found multiple contracts in the same crate");
+    }
+
+    let (module_id, file_id) = contract_module_file_ids.pop().unwrap();
+
+    // If compute_note_hash_and_nullifier is already defined by the user, we skip autogeneration in order to provide an
+    // escape hatch for this mechanism.
```

Comment:
Can you put a TODO here linking to this issue: https://github.com/AztecProtocol/aztec-packages/issues/4647